src/runtime/runtime_data.cs

```diff
@@ -0,0 +1,426 @@
+
+            foreach (var item in objs)
+            {
+                item.Value.ExecutePostActions();
+                XDecref(item.Key.pyHandle);
+            }
+            foreach (var item in clsObjs)
+            {
+                item.Value.ExecutePostActions();
+            }
+        }
+
+        public static bool HasStashData()
+        {
+            return PySys_GetObject("clr_data") != IntPtr.Zero;
+        }
+
+        public static void ClearStash()
+        {
+            PySys_SetObject("clr_data", IntPtr.Zero);
+        }
+
+        private static void StashPushObjects(RuntimeDataStorage storage)
+        {
+            var objs = ManagedType.GetManagedObjects();
+            var extensionObjs = new List<ManagedType>();
+            var wrappers = new Dictionary<object, List<CLRObject>>();
+            var serializeObjs = new CLRWrapperCollection();
+            var contexts = new Dictionary<IntPtr, PyObjectSerializeContext>();
+            foreach (var entry in objs)
+            {
+                var obj = entry.Key;
+                XIncref(obj.pyHandle);
+                switch (entry.Value)
+                {
+                    case ManagedType.TrackTypes.Extension:
+                        Debug.Assert(obj.GetType().IsSerializable);
+                        var context = new PyObjectSerializeContext();
+                        contexts[obj.pyHandle] = context;
+                        obj.Save(context);
+                        extensionObjs.Add(obj);
+                        break;
+                    case ManagedType.TrackTypes.Wrapper:
+                        // Wrapper must be the CLRObject
+                        var clrObj = (CLRObject)obj;
+                        object inst = clrObj.inst;
+                        CLRMappedItem item;
+                        List<CLRObject> mappedObjs;
+                        if (!serializeObjs.TryGetValue(inst, out item))
+                        {
+                            item = new CLRMappedItem(inst)
+                            {
+                                Handles = new List<IntPtr>()
+                            };
+                            serializeObjs.Add(item);
+
+                            Debug.Assert(!wrappers.ContainsKey(inst));
+                            mappedObjs = new List<CLRObject>();
+                            wrappers.Add(inst, mappedObjs);
+                        }
+                        else
+                        {
+                            mappedObjs = wrappers[inst];
+                        }
+                        item.Handles.Add(clrObj.pyHandle);
+                        mappedObjs.Add(clrObj);
+                        break;
+                    default:
+                        break;
+                }
+            }
+
+            var wrapperStorage = new RuntimeDataStorage();
+            WrappersStorer?.Store(serializeObjs, wrapperStorage);
```

Comment:
If `WrappersStorer` is `null`, won't this data be simply lost?

Reply:
Yes. The runtime can not decide what's the serialize way for the user objects.